pkg/virt-controller/watch/vmi.go

```diff
@@ -2399,3 +2401,97 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
+	Status  k8sv1.ConditionStatus
+	Reason  []string
+	Message []string
+}
+
+func (c *VMIController) aggregateDataVolumesConditions(vmiCopy *virtv1.VirtualMachineInstance, dvs []*cdiv1.DataVolume) {
+	if len(dvs) == 0 {
+		return
+	}
+	const ConditionUninitialized k8sv1.ConditionStatus = "Uninitialized"
+
+	allDVConditions := map[cdiv1.DataVolumeConditionType]aggregateDataVolumeCondition{
+		cdiv1.DataVolumeReady: {
+			Status: ConditionUninitialized,
+		},
+		cdiv1.DataVolumeBound: {
+			Status: ConditionUninitialized,
+		},
+		cdiv1.DataVolumeRunning: {
+			Status: ConditionUninitialized,
+		},
+	}
+
+	for _, dv := range dvs {
+		for _, condition := range dv.Status.Conditions {
+			if allDVConditions[condition.Type].Status == ConditionUninitialized {
+				allDVConditions[condition.Type] = aggregateDataVolumeCondition{
+					Status:  condition.Status,
+					Reason:  []string{condition.Reason},
+					Message: []string{condition.Message},
+				}
+			} else {
+				if condition.Type == cdiv1.DataVolumeRunning {
+					aggregateCondsTrueWins(condition, allDVConditions)
+				} else {
+					aggregateCondsFalseWins(condition, allDVConditions)
+				}
+			}
+		}
+	}
+
+	dvsReadyCondition := virtv1.VirtualMachineInstanceCondition{
+		Type:   virtv1.VirtualMachineInstanceDataVolumesReady,
+		Status: k8sv1.ConditionUnknown,
+	}
+
+	if allDVConditions[cdiv1.DataVolumeBound].Status == k8sv1.ConditionFalse {
+		dvsReadyCondition.Status = k8sv1.ConditionFalse
+		dvsReadyCondition.Reason = virtv1.VirtualMachineInstanceReasonNotAllDVsBound
+		dvsReadyCondition.Message = "Not all of the VMI's DVs are bound: " + strings.Join(allDVConditions[cdiv1.DataVolumeBound].Message, ",")
+	} else if allDVConditions[cdiv1.DataVolumeRunning].Status == k8sv1.ConditionTrue {
+		dvsReadyCondition.Status = k8sv1.ConditionFalse
+		dvsReadyCondition.Reason = virtv1.VirtualMachineInstanceReasonDVRunning
+		dvsReadyCondition.Message = "At least one of the VMI's DVs is still running (import/upload/clone): " + strings.Join(allDVConditions[cdiv1.DataVolumeRunning].Message, ",")
+	} else if allDVConditions[cdiv1.DataVolumeReady].Status == k8sv1.ConditionTrue {
+		dvsReadyCondition.Status = k8sv1.ConditionTrue
+		dvsReadyCondition.Reason = virtv1.VirtualMachineInstanceReasonAllDVsReady
+		dvsReadyCondition.Message = "All of the VMI's DVs are bound and not running"
+	}
+
+	vmiConditions := controller.NewVirtualMachineInstanceConditionManager()
+	vmiConditions.UpdateCondition(vmiCopy, &dvsReadyCondition)
+
+}
+
+func aggregateCondsFalseWins(condition cdiv1.DataVolumeCondition, allDVConditions map[cdiv1.DataVolumeConditionType]aggregateDataVolumeCondition) {
+	if condition.Status == k8sv1.ConditionFalse || condition.Status == k8sv1.ConditionUnknown {
+		var status k8sv1.ConditionStatus
+		if allDVConditions[condition.Type].Status == k8sv1.ConditionTrue {
+			status = condition.Status
+		}
+		allDVConditions[condition.Type] = aggregateDataVolumeCondition{
+			Status:  status,
```

Comment:
`status` is uninitialized when `allDVConditions[condition.Type].Status != k8sv1.ConditionTrue`?

Reply:
Right, this was somehow masked by the set of if-s starting on line 2452 but let's explicitly avoid it with a safe initialization.
Let's also enlarge the test matrix in the unit tests.